[2단계 - 주문 기능 구현] 도이(유도영) 미션 제출합니다.

yoondgu opened the pull request:
안녕하세요 라빈! 🍨 리뷰이 도이입니다!! 
마지막 미션을 함께 하게 되어 영광이에요~ 
요청이 많이 늦었지만.. 같이 많이 대화하며 배워가고 싶습니다 😆  잘 부탁드립니다!! ㅎㅎ

백엔드, 프론트엔드 페어들과 협업하여 이번 미션 진행하였습니다.

- API 명세
    - api  문서 자동화는 아직 하지 못해서 대신 협업하며 사용한 명세 문서를 [API.md](https://github.com/woowacourse/jwp-shopping-order/blob/f141cfdffb5a82c2addd3bee6121a47a70641f43/API.md)로 먼저 작성해두었습니다.
- DB 환경
    - 생각보다 구현에 시간이 오래 걸려, DB도 AWS 서버에 띄우는 작업은 추가로 월요일에 진행 예정입니다. 지금은 H2 인메모리 DB를 사용 중입니다! (코드에는 영향이 없을 것 같아 리뷰 요청부터 드립니다..!) 
- 추가 구현 기능
   -  [README.md](https://github.com/woowacourse/jwp-shopping-order/blob/f141cfdffb5a82c2addd3bee6121a47a70641f43/README.md)의 기능 목록에서 보시다시피 '쿠폰'을 적용하였습니다.
   - 협업 회의 결과, 사용자가 쿠폰을 가지고 있다는 전제 하에 기능을 구현하는 것으로 진행했습니다. (쿠폰 관리자 기능은 백엔드 측에서 각자 원하면 진행)
- [리뷰 범위](https://github.com/woowacourse/jwp-shopping-order/pull/47/files/ff02a133611e306138e2e2b3ee79074ee192ddc3..f141cfdffb5a82c2addd3bee6121a47a70641f43) 

## 개발 프로세스

중간에 요구사항이 추가되는 경험을 하고 싶어서 아래와 같은 순서로 진행했습니다.

`기본 기능 팀 회의(API 명세 작성) - 기본 기능 구현 - 추가 기능 팀 회의(API 명세 변경 및 작성) - 추가 기능 구현` 

추가 기능에 따른 코드 변경이 여러 도메인에 엮여있어서
추가 기능 구현에서는 커밋 단위도 상당히 커졌고, 쉬운 일이 아니라는 걸 느낄 수 있었습니다. ㅎㅎ
그 과정에서 배포한 뒤 버그를 발견한 적도 많고요.. !! 
앞으로는 구현 내용이 복잡할 수록 TODO를 더 작게 쪼개서 침착하게 진행해야겠다는 교훈을 얻고 있습니다.

## 중점을 둔 부분 / 고민거리

구현하면서 중점을 둔 부분은 아래와 같습니다!

### ATDD

- 고려해야 할 사항이 많은 만큼 개발 과정에 길을 덜 잃고 싶었어요. 그래서 인수 테스트를 작성하고, 해당하는 기능을 구현하는 순서로 진행하고자 했습니다.
- 제가 작성한 코드가 인수 테스트라고 생각한 이유
    - “주문 삭제”에 대한 테스트를 예로 들자면, 주문 추가, 주문 삭제, 주문 조회(삭제 확인)를 하는 사용자 시나리오를 더미 데이터나 다른 코드에서 DB에 접근하는 등의 방식을 사용하지 않고 모두 테스트 컨텍스트를 사용해 실제 요청-응답으로 실행하기 때문입니다.

(하지만 한 테스트 클래스가 너무 크고,  다른 종류의 테스트들은 너무 미진하네요.. 추가하도록 하겠습니다!)

### auto increment 값에 의존적이지 않은 테스트 코드 작성

- 테스트 데이터 초기화가 까다로워지는 이유 중 하나가 auto increment 값도 초기화해야 한다는 점이라고 생각하는데요. 복잡한 데이터 초기화 설정에 의지하면, 설정을 조금만 바꿔도 테스트가 쉽게 터져서 힘들었던 기억이 있습니다.
- 최대한 제어하기 어려운 요소들을 배제하며 테스트를 작성하고 싶었습니다. 그래서 이번에는 최대한 테스트 코드가 auto increment 값에 의존하지 않도록 구현해보려고 했습니다.

### 도메인 객체 중심으로 설계하기

- 테이블 간 참조가 많은 도메인들에 대하여, DB join 대신 Repository 계층에서 여러 Dao를 사용해 직접 객체를 조립해 주는 방식으로 진행했습니다.
- 이전 미션들에서는 항상 DB join 방식만 사용했어서, 이번에는 완전 반대 방식으로 극단을 경험해보고 싶었습니다! 그러다보니 ProductRepository와 같이 불필요한 계층 분리도 있지만, 설계의 일관성을 가져가고 싶었습니다.
- 이 과정에서 제가 느낀 장단점은 아래와 같습니다.
    - 장점
        1. 도메인 객체는 자기 자신의 id 이외에는, id가 아닌 필드로 다른 도메인 객체를 참조합니다. 따라서 도메인 간 비즈니스 로직을 조금이나마 객체지향적으로 풀 수 있었습니다. 
        `ex` Order를 취소할 때, 해당 객체가 가지고 있는 Coupon을 환불함
        2. Dao에서는 재사용성이 높은 메서드를 제공하고, 조립을 Repository에서 하기 때문에 SQL 수정을 해야 하는 일이 적습니다. 물론 대신 SQL 외의 코드를 수정해야겠지만, SQL은 벤더사 종속적이고 더 실수하기 쉽기 때문에 이를 줄이는 편이 더 좋게 느껴집니다.
    - 단점
        1. DB join으로는 한번에 조회할 수 있는 정보를 여러 번의 쿼리를 날려야 합니다. 직접 느꼈다보기다는, 처음에는 이 방법을 사용하지 않은 이유입니다. 하지만 지금은 성능 이슈가 가장 중요한 상황이 아니라고 생각했습니다.
        2. Order도 있고, OrderEntity도 있고, 변환 로직도 필요하고 … 복잡도가 올라간 만큼 필드 추가와 같은 조금의 변경사항에도 수정할 코드 양이 많아 힘들었습니다.

### 불가피한 Null값?

- 이 부분은 아직 진행중인 고민입니다!
- 이번 미션은 유독 null을 직접 쓰거나, NPE를 마주하는 일이 많았습니다. 객체가 null이 될 수 있는 필드를 가지는 상황들 때문이었습니다.
    1. 객체를 영속화하는 방식에서 null
        - OrderRepository.add(Order order)에 전달할 Order 객체는 아직 id, createdAt 값을 가지고 있지 않다.
    2. 도메인 로직에서의 null
        - Order 객체는 쿠폰을 사용하지 않을 수도 있다.
- 이렇게 어플리케이션 전반적으로 Null이 여기 저기 돌아다닐 수 있는 상황인 것 같아서 여전히 찜찜합니다..
- 가능한 Optional을 활용해보려고 했지만, 언제 어떤 기준으로 어디서 Optional을 주고, 어디서 null을 꺼내야 할지 아직 기준이 잘 서지 않습니다..! 혹시 라빈은 어떤 방식으로 접근하시는지 여쭤봐도 될까요?

yoondgu committed:
feat: Rest API 모든 Origin 허용

yoondgu committed:
refactor: CORS 설정 Config 전역 설정으로 변경

yoondgu committed:
fix: CORS 설정 Location Header 허용

yoondgu committed:
fix: CORS 설정 PATCH 메서드 추가

yoondgu committed:
fix: CORS 설정 OPTIONS 메서드명 수정

yoondgu committed:
refactor: CORS 설정 HttpMethod 이넘 사용

yoondgu committed:
feat: 상품 목데이터 추가

yoondgu committed:
fix: 상품 목데이터 추가 SQL 컬럼명 수정

yoondgu committed:
feat: CORS 접근 Origin 제한

yoondgu committed:
feat: 장바구니 담기 및 수량 변경 검증 로직 추가

yoondgu committed:
feat: 장바구니 여러 개 삭제하도록 API 명세 및 기능 변경

yoondgu committed:
fix: CORS URL 임시로 모두 열어줌

yoondgu committed:
fix: CORS URL 임시로 모두 열어줌

yoondgu committed:
fix: CORS 허용 Origin 수정

yoondgu committed:
feat: 주문 기능 관련 DB DDL 추가

yoondgu committed:
docs: 2단계 요구사항 작성

yoondgu committed:
refactor: 상품 도메인 객체 금액 원시값 포장, 데이터타입 변경

yoondgu committed:
test: 사용하지 않는 테스트 삭제

yoondgu committed:
refactor: 금액 원시값 포장 반영, 코드 포맷 수정

yoondgu committed:
feat: 주문 추가 기능 구현

yoondgu committed:
refactor: 사용자 인증 관련 패키지, 메서드 및 클래스 분리

yoondgu committed:
refactor: 프로젝트 코드 전체 final 키워드 적용

yoondgu committed:
style: 코드 컨벤션 통일, 사용하지 않는 import문 삭제

yoondgu committed:
refactor: 사용자 관련 뼈대코드 리팩터링

- 사용하지 않는 메서드 삭제
- RowMapper 객체 매번 생성하는 대신 상수로 관리
- deprecated 된 Dao 메서드 수정, try catch 처리

yoondgu committed:
refactor: 장바구니 관련 뼈대코드 리팩터링

- SQL 줄바꿈 수정
- Row Mapper 불필요한 변수 할당 제거
- SimpleJdbcInsert 활용

yoondgu committed:
refactor: 커스텀 예외에 대한 Exception Handler 메서드 추가

yoondgu committed:
test: 주문정보 추가 요청 실패 테스트 추가

yoondgu committed:
feat: 주문 목록 조회 기능 구현

yoondgu committed:
feat: 주문 추가 기능 명세 변경 반영, 검증로직 추가

yoondgu committed:
fix: 주문 추가 기능 장바구니 삭제하도록 수정

yoondgu committed:
feat: 주문 상세 조회 기능 구현

yoondgu committed:
feat: 주문 취소 기능 구현

yoondgu committed:
fix: Dto 프론트엔드 변수명 매핑 오기 수정

yoondgu committed:
chore: 프로덕션 코드 장바구니, 주문 더미 데이터 추가

yoondgu committed:
fix: CORS 허용 Origin 수정

yoondgu committed:
fix: 장바구니 조회 버그 수정

- 상품 아이디로 장바구니 아이디를 가져오는 버그 수정

yoondgu committed:
fix: 502 에러로 인한 CORS ORIGIN 설정 복구

yoondgu committed:
fix: 502 에러 해결 후 CORS ORIGIN 재설정

yoondgu committed:
refactor: 주문, 주문 상품목록 DAO 분리

- 실행 쿼리 수를 줄여 성능 최적화하기 위해 JOIN을 사용했으나
- 현재로서는 불필요한 최적화보다는 역할 책임의 분리가 중요하다고 생각하여 변경함.
- 서비스 계층에 많아진 Entity 간 조립 로직을 별도 계층으로 분리 예정

yoondgu committed:
refactor: 주문, 주문 상품목록 Repository 계층 도입

- 다른 도메인에도 도입 예정

yoondgu committed:
feat: 누락된 예외 헨들링 추가

yoondgu committed:
feat: Repository 계층 도입

- 테이블 간 참조 관계가 복잡한 객체는 Repository 계층에서 DB 조회 결과를 조립하도록 함.
- 상품(Product), 회원(Member)는 다른 테이블을 참조하는 관계가 없지만 설계의 일관성을 위해 Repsitory 계층 도입함.

yoondgu committed:
docs: 협업 회의 후 추가 요구사항 명세 업데이트

yoondgu committed:
feat: 주문 상태 정보 추가, 주문 취소 기능 구현

- 기존 주문 취소 기능은 주문 정보 삭제 기능임.
- 주문 상태("결제완료", "결제취소") 정보를 추가하고, 상태를 변경하는 API 명세 추가함.

yoondgu committed:
feat: 쿠폰 도메인 추가에 따른 기능, API 명세 반영

- 주문 추가 시 쿠폰 1장을 적용할 수 있음
- 쿠폰 관련 테이블 추가
- 주문 목록 조회에 결제 총 금액, 주문일, 결제 상태 정보 추가
- 주문 상세 조회에 결제 총 금액, 쿠폰, 주문일 정보 추가
- 주문 삭제/취소 시 쿠폰 반영 로직은 미반영 상태임

yoondgu committed:
feat: 주문 추가/취소 시 쿠폰 사용 상태 변경 기능 구현

- 주문을 추가 / 취소에 따라 쿠폰 사용 상태를 변경하도록 구현함
- 쿠폰 테이블 단순화함

yoondgu committed:
feat: 사용자 쿠폰 목록 조회 기능 구현

yoondgu committed:
fix: 주문 시 배송비 대신 상품 총 금액 반영되는 버그 수정

yoondgu committed:
refactor: 명시가 필요한 테스트 값 변수 처리

yoondgu committed:
refactor: 쿠폰 미적용 주문에 대한 null 처리 방식 변경

- 쿠폰을 적용하지 않은 주문 조회 시 coupon DTO를 null로 가짐
- 이에 대한 null 할당을 CouponResponse 생성 시 하는 것이 아니라, OrderDetailResponse 생성 시 하도록 함
- CouponResponse가 쿠폰 목록 조회에 쓰일 때는 null이 되지 않도록 함

yoondgu committed:
refactor: 주문 기능 인수 테스트 검증 내용 추가

- 테스트 시 주문 기능에 대한 실제 결과 값을 확인하는 내용까지 추가함

yoondgu committed:
refactor: 테스트 코드 모호한 변수명 수정

- 인수 테스트에서 여러 개의 response 를 다룰 때 사용하는 변수명 개선
- 자원을 생성하는 요청에 대한 응답: createdResponse
- 자원을 조회하는 요청에 대한 응답: foundResponse

yoondgu committed:
fix: 쿠폰 API CORS 설정

yoondgu committed:
docs: 기능 목록 구현 항목 업데이트

yoondgu committed:
build: DDL 프로덕션과 테스트 설정 분리

- 실제 배포 시 이미 테이블이 있으면 다시 생성하지 않도록 함
- 테스트 컨텍스트 생성 시 항상 테이블을 새로 만들도록 함

yoondgu committed:
style: 불필요한 주석 삭제

yoondgu committed:
docs: API 명세서 추가

yoondgu committed:
docs: API 명세서 스타일 수정

yoondgu committed:
docs: API 명세서 마크다운 테이블 오류 수정

yoondgu committed:
fix: 쿠폰 목록 조회 버그 수정

- 잘못된 더미 데이터 DML 수정
- CouponType 조회 코드 실수 수정

yoondgu committed:
docs: API.md 테이블 깨짐 수정

yoondgu committed:
feat: 주문 상세 정보 조회 API 명세 변경 반영

yoondgu committed:
chore: 쿠폰 더미데이터 대량 추가

- 쿠폰 발급 기능 미구현으로 인해 편의를 위해 발급 쿠폰을 대량으로 추가함

yoondgu committed:
fix: 주문 시 쿠폰 반영 버그 수정

- f141cf 커밋과 동일한 코드 실수로 인한 버그 수정
- coupon type 을 coupon id로 조회하는 버그

yoondgu committed:
style: sql 파일 코드 정렬 수정

yoondgu committed:
fix: 주문 목록 조회 시 총금액 계산 오류 수정

- 주문 목록 조회 시 쿠폰 정보 조회하지 않아 총 금액에 쿠폰 반영 누락되는 문제 해결

yoondgu commented:
코드 리뷰 요청 후 버그를 발견 / 팀 API 명세 변경 더미데이터 요청으로 추가 push가 존재합니다..!
git repository에서 가져와 배포하는 방식으로 되어있어 불가피하게 push를 하게 되었습니다 😢 죄송합니다.
리뷰 요청드릴 때 지정했던 커밋 범위 [f141cfd](https://github.com/woowacourse/jwp-shopping-order/pull/47/commits/f141cfdffb5a82c2addd3bee6121a47a70641f43)까지만 봐주셔도 정말 감사하겠습니다..!

yoondgu committed:
feat: 프론트엔드 CORS 허용 Origin 추가

giantim commented on src/main/java/cart/ui/CouponApiController.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<CouponResponse>> showCoupons(final Member member) {
+        return ResponseEntity.ok(couponService.findByMember(member));
```
DTO 없이 도메인 객체를 직접 API 통신에 사용한다면 변경에 취약한 도메인으로 발전할거 같은데 DTO를 사용하지 않으신 이유가 있으실까요?

yoondgu replied:
이 어플리케이션에서 Member에 대한 다른 기능(사용자 관리, 사용자 정보 수정 등)은 존재하지 않아서
도메인, DTO, Entity 등을 구분하지 않고 단순하게 하나의 객체 Member만 사용했습니다.

Member에 아이디, 비밀번호 정보를 가지고 있어 개인정보가 유출 될 수 있는 문제는 있다고 생각됩니다.
그래서 ArgumentResolver에서 Member를 반환하는 대신, member의 Id만 반환하는 것은 고려할 수 있을 것 같습니다.

하지만 "API 통신에 사용한다면 변경에 취약한 도메인으로 발전"한다는 게 정확히 상상이 가지 않습니다.. !ㅜㅜ
혹시 어떤 경우인지 조금만 더 설명해주실 수 있을까요?

yoondgu replied:
디엠으로 이해하기 쉽게 설명해주셔서 감사합니다!! 😀
말씀해주신 것처럼, Member라는 도메인에 대한 새로운 요구사항이 생겨 변경이 일어났을 때를 고려하지 못했던 것 같아요!
그러고보니 Order, CartItem, Coupon이 Member를 가지고 있는 것이 변경에 취약하고 불필요한 의존으로 보이더라구요.
그래서 이 의존을 끊기 위해서, memberId만을 전달하도록 전면 수정했습니다. ([9bfbc9e](https://github.com/woowacourse/jwp-shopping-order/pull/47/commits/9bfbc9ee099068a8503d95dfb92e22298edc3768))

Entity에서는 DB 설계상 id로 참조하고 있다고 해도, 도메인 또한 Member를 가지고 있을 필요는 없어보였습니다.
회원 유효성을 확인해야 하는 부분은 서비스 계층에서 memberId만 가지고 수행해도 충분하기 때문입니다.
따라서 Member를 전달하고 굳이 id만 가지고 있는 Member를 참조하도록 했던 로직들을 모두 memberId를 주고 받도록 수정하였습니다.

`로그인 사용자의 Order, CartItem, Coupon인지 확인하는 것`은 `findByIdForMember(memberId, id)` 와 같이 쿼리 조회에서 한번에 검증하도록 하였습니다.

다만 그러다보니 
Order, Coupon 간의 관계를 예로 들자면
기존에는 OrderRepository -> CouponRepository의 의존관계만 존재했는데
OrderService -> CouponService의 의존관계가 동시에 존재하게 되었습니다.
`로그인 사용자의 Order, CartItem, Coupon인지 확인` 하는 로직을 각 서비스에서 담당하게 했기 때문입니다.
의존의 흐름은 같은 방향이긴 하지만,
또 OrderService 또한 CouponRepository를 의존하는 것보다는 계층 분리를 하는 게 좋다고 생각하지만,
그럼에도 각 계층에서 계속 또 도메인 간 의존을 해야 한다는게 불필요하게 중첩된 의존 관계 같아서 고민되는 지점입니다..!

giantim replied:
말씀주신 방향대로 서비스 클래스간의 의존 방향은 항상 신경써야 하는 부분이에요. 순환 의존성이 발생할 수 있고 설계한 도메인간의 의존 방향을 따르지 않는 실수를 할 가능성이 있기 때문이에요.
저는 지금 구조에서는 OrderService에서 CouponRepository를 직접 의존해도 괜찮을것 같네요!

giantim commented on src/main/java/cart/repository/CouponRepository.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private CouponTypeEntity findCouponType(final Long id) {
+        final CouponTypeEntity couponType = couponTypeDao.findById(id)
+                .orElseThrow(() -> new IllegalStateException("illegal data exists in table COUPON; coupon_type_id"));
+        return couponType;
```
```suggestion
        return couponTypeDao.findById(id)
                .orElseThrow(() -> new IllegalStateException("illegal data exists in table COUPON; coupon_type_id"));
```
위 처럼 간결하게 바꿔보면 어떨까요? 😉

yoondgu replied:
앗 ! 꼼꼼하게 봐주셔서 감사합니다 ㅎㅎ 반영하겠습니다~!

giantim commented on src/main/java/cart/infrastructure/AuthInfo.java:
```diff
@@ -0,0 +1,8 @@
+package cart.infrastructure;
```
음.. 패키지 이름이 모호해서 정말 다양한 클래스들이 모일것 같아요.
인프라의 정의를 어디까지 할 것인가에 대해 계속 의문이 드는데요. 도이가 생각하신 인프라의 정의가 어떻게 되실까요? 인증에 관한 기능들이 어떤 이유로 인프라에 포함이되고 설정 파일들은 왜 인프라에 들어가야 하는지 의문이네요 🤔
에외 처리 핸들러는 지금 패키지보다 exception 패키지가 더 어울릴 수 있을것 같고, 설정파일들의 경우 저는 관리의 편의성을 위해 Application 클래스와 같은 레벨에 정의하는데요. 해당 패키지에 대한 명확한 역할을 다시 생각해보면 좋을것 같아요!

yoondgu replied:
앗 그러게요.. 패키지 분리에 대해 깊게 생각하지 못하고 진행한 것 같아요..!!
보다 구체적인 패키지명으로 변경해주도록 하겠습니다.
ExceptionHandler는 exception으로 옮기고,
Infrastructure 대신 config라는 패키지를 두어서
어떤 종류의 config냐에 따라 내부 패키지를 나눠주도록 하겠습니다.

giantim commented on src/main/java/cart/domain/Member.java:
```diff
@@ -1,16 +1,29 @@
+        this(id, null, null);
+    }
+
+    public static Member from(final MemberEntity member) {
```
Member  클래스는 우리가 정의한 도메인 객체고 MemberEntity는 영속화된 데이터를 매핑하는 객체인거 같은데요.
그럼 도메인 객체를 생성하는 책임은 둘 중 어느 클래스가 갖는게 맞을까요? 외부 기술에 따라 변경이 발생할 수 있는 클래스를 사용하여 도메인에서 직접 생성하는 방식이 적합할지 고민이네요 🤔

yoondgu replied:
동의합니다..!
MemberEntity는 DB 테이블의 구성에 따라 변경이 발생할 수 있습니다.
그런데 지금은 Member에서 MemberEntity의 값을 꺼내 객체를 만듬으로써 MemberEntity의 필드 구성에 의존하게 된 것 같습니다.
따라서 Entity 객체들이 도메인 객체를 생성하는 방식으로 수정하겠습니다!
생각지 못한 부분에 대해 고민해볼 수 있었습니다. 감사합니다 ㅎㅎ

giantim commented on src/main/java/cart/dao/entity/MemberEntity.java:
```diff
@@ -0,0 +1,29 @@
+        return password;
+    }
+
+    public boolean checkPassword(final String password) {
+        return this.password.equals(password);
+    }
```
사용하지 않는 메소드는 지워주세요!

yoondgu replied:
반영하겠습니다!

giantim reviewed (changes requested):
도이 안녕하세요! 리뷰어 라빈입니다 👋
이번 미션 진행하시면서 실제와 같은 개발 프로세스 경험을 위해 직접 제약사항도 설정해보시고 좋은 경험 해보셨을것 같네요 👍 그 과정에서 설계에 대해 어떻게 접근할지에 대해 느낀점도 있고 좋은 학습방향인거 같아요! 그리고 도메인 중심적으로 설계하셔서 요구사항에 대한 기능들이 도메인 클래스에 잘 위치하고 있는것 같습니다! 
불가피한 null에 대해서는 충분히 공감이 되는 내용이네요. 영속화 하는 시점에 값을 정의하지 못할 수 있는 경우가 가장 많은것 같아요. 이럴때 optional을 사용할지도 고민이 되시는것 같은데요. 저는 필드가 wrapper / primitive 타입을 이용해 nullable을 나타내는것 같아요. 기본값으로 대체해도 되는 값의 경우 primitive 타입을 사용할 수 있을것 같고 그게 아니라 nullable 하게 지정해야 하는 값들에 대해서는 wrapper 타입을 사용해봐도 좋을것 같네요!
리뷰 몇 가지 남겨드렸는데 같이 의논해보면 좋을것 같네요! 😆

yoondgu committed:
docs: 피드백 반영 및 리팩터링 목록 작성

yoondgu committed:
refactor: 사용하지 않는 메소드 삭제

yoondgu committed:
refactor: 모호한 패키지 구조 변경, 중복 config 파일 통합

yoondgu committed:
refactor: Dao Entity에서 도메인 객체 생성하도록 책임 변경

yoondgu committed:
refactor: 변경에 취약한 도메인 의존 끊기

- 다른 도메인 객체가 member를 의존하지 않도록 변경

yoondgu committed:
refactor: 사용자 정보 없으면 null 반환 대신 예외 발생

yoondgu committed:
refactor: Dao 테스트 추가 작성

yoondgu committed:
refactor: 도메인 단위 테스트 추가 작성

yoondgu committed:
style: 테스트 메서드 displayName 누락 추가

yoondgu committed:
feat: API 문서 자동화 swagger 기본 설정

yoondgu committed:
refactor: api 문서 리소스 별 분리, content type 지정

yoondgu committed:
refactor: api 문서 리소스 별 분리, 세부 설정 전역 처리

- 리소스 별 분리
- content type 설정
- basic auth header 설정
- argumentResolver 로 획득하는 값은 파라미터로 인식하지 않도록 설정

yoondgu commented:
라빈 남겨주신 리뷰 덕분에 놓쳤던 부분들을 고민하며 피드백 반영할 수 있었습니다! 감사합니다 ㅎㅎ

- 피드백 반영 목록은 [README](https://github.com/woowacourse/jwp-shopping-order/blob/e74d6f8fce5c92882c2c330ab1193b51369bf3df/README.md)에 코멘트 링크와 함께 정리하였습니다.
   - 디엠으로 질문 드렸던 Member와 도메인 변경 관련 내용에도 [추가 답변](https://github.com/woowacourse/jwp-shopping-order/pull/47#discussion_r1223316503) 남겼습니다!
   - Null 관련 답변도 정말 감사합니다! 말씀해주신 내용을 생각하며 제 코드를 다시 보니 id에 대한 Wrapper, Primitive 타입을 정말 일관성 없이 쓰고 있었더라구요.. 🥲 그래서 null이 아니면 안되는 파라미터에서는 꼭 Primitive를 써서 좀 더 빠르게 예외를 던지도록 수정했습니다! 그리고 좀 더 Optional에서 지원하는 기능들을 효율적으로 써보려고 했어요.
   - 별도 서버 MySQL 적용은 실제 배포 버전에서는 적용된 상태입니다.
   (external properties 설정 대신 vim으로 수정해서 배포했습니다 🥹)
   - API 문서 자동화의 경우, RestDocs와 Swagger 중 Swagger를 선택한 이유
      - RestDocs가 테스트 기반이고 프로덕션 코드에 영향을 주지 않는다는 장점이 더 커보였으나, 현재 제 상황에서는 이를 적용하기 위해 기존에 작성한 인수 테스트와는 별개로 MockMvc로 컨트롤러 테스트를 작성해야 했습니다.
      - 만약 프로젝트 시작 단계에서 이를 적용했다면, 테스트 작성과 함께 문서 자동화를 할 수 있으니 좋았겠지만 지금 RestDocs만을 위해 테스트 코드를 더 작성하는 것은 비효율적이라 판단했습니다.
      - 따라서 아쉽지만 지금 시점에서는 빠르게 적용 가능한 Swagger가 낫다고 생각했고, 그 대신 최대한 전역 설정으로 풀어보려고 노력해봤습니다!
   
지난 리뷰 요청 때 공유드리지 못했던 배포 주소, api 문서 주소도 함께 남겨둡니다..ㅎㅎ

### 프론드엔드 배포 URL
크루 | URL
-- | --
야미 | https://feb-dain.github.io/react-shopping-cart-prod
코난 | https://cruelladevil.github.io/react-shopping-cart-prod

### REST API 배포 도메인, 설정 URL
https://wooteco.store
설정 기능 | URL
-- | --
사용자 선택 | https://wooteco.store/settings
상품 관리자 | https://wooteco.store/admin

### REST API 자동화 문서
https://wooteco.store/swagger-ui/index.html

이번 리뷰도 잘 부탁드립니다!!! 🙇‍♀️🙇‍♀️

giantim commented on src/main/java/cart/repository/CouponRepository.java:
```diff
@@ -23,30 +23,37 @@ public CouponRepository(final CouponDao couponDao, final CouponTypeDao couponTyp
 
-    public void updateStatus(final Coupon coupon) {
-        couponDao.updateStatus(CouponEntity.from(coupon));
+    public Optional<Coupon> findById(final long id) {
+        if (Objects.isNull(id)) {
+            return Optional.empty();
+        }
```
현재 CouponRepository에서 매개변수로 받는 id 값들은 long 타입인데요. primitive 타입이어서 Objects.isNull() 검사 수행시 항상 false를 반환할것 같아요. 해당 구문을 없애도 무방해 보입니다!

giantim reviewed (approved):
도이 안녕하세요! 라빈입니다 👋
기존 구조도 굉장히 잘 구성해주셨었는데 리뷰 이후 리팩토링 해주신 부분들까지 합쳐져서 더 보기 좋아진 코드가 된것 같네요 👍 개발 프로세스 과정과 배포 내용들까지 상세하고 꼼꼼하게 작성해 공유해주셔서 기능적으로 이해하는데 많은 도움이 되었습니다 😉
질문 주신 내용에 추가 코멘트와 한가지 간단한 리뷰 남겨드렸어요! 다음 단계 미션 진행하시면서 고민해보시면 좋을것 같고 이번 단계 진행하시느라 수고 많으셨습니다 👋
